egress-gateway: Add generic --enable-egress-gateway flag

syedazeez337 opened the pull request:
# Add generic --enable-egress-gateway flag

## Description

This PR adds a new `--enable-egress-gateway` flag that enables egress gateway functionality for both IPv4 and IPv6. The existing `--enable-ipv4-egress-gateway` flag is marked as deprecated but still functional for backward compatibility.

## Changes

- Added new `EnableEgressGateway` flag in `pkg/option/config.go`
- Updated the DaemonConfig struct to include the new flag
- Added the new flag to the CLI options in `daemon/cmd/daemon_main.go`
- Updated the egressgateway manager to check for both flags
- Updated the policy map creation to check for both flags
- Updated the CRD resource names to check for both flags
- Added the new flag to the operator flags

## Testing

- Ran unit tests for the affected packages
- Created a simple test program to verify the new flag is properly defined and accessible
- Verified that both flags can be set in the DaemonConfig struct

## Related Issues

Fixes #38957
Fixes #38962

syedazeez337 committed:
egressgateway: Add IPv6 support

This commit adds IPv6 support to the egress gateway feature:

- Add a new --enable-egress-gateway flag that enables egress gateway for both IPv4 and IPv6
- Mark the existing --enable-ipv4-egress-gateway flag as deprecated
- Update the validation format for EgressIP from ipv4 to ip to support both IPv4 and IPv6 addresses
- Modify the code to handle IPv6 egress IPs
- Add IPv6 constants for excluded CIDRs, gateway not found, and egress IP not found
- Fix a bug in the updateEgressRules6 function where it was using IPv4 constants for IPv6 excluded CIDRs
- Update the Helm chart to use the new flag
- Update documentation to reflect the new IPv6 support and add examples for IPv6 egress gateway policies
- Update metrics and cilium-cli feature detection to check for both flags
- Update BPF code to use both flags

Signed-off-by: Syed Azeez <syedazeez337@gmail.com>

joestringer commented on .github/release-notes/egress-gateway-flag.yaml:
Please drop the changes to `.github/release-notes`. Instead, you can create a section in the PR description in this form and the change will show up in the release notes:

````
```release-note
Add a new `--enable-egress-gateway` flag that enables egress gateway for both IPv4 and IPv6, deprecating the IPv4-only `--enable-ipv4-egress-gateway` flag.
```
````

joestringer commented on Documentation/network/egress-gateway/egress-gateway.rst:
```diff
@@ -80,7 +80,7 @@ by an egress gateway policy must be in the same cluster as the selected pods.
 
-Egress gateway is not supported for IPv6 traffic.
+Egress gateway is supported for both IPv4 and IPv6 traffic.
 
```
No need to have an item under "Incompatibility" that says all expected traffic types are supported.
```suggestion
```

joestringer commented on Documentation/network/egress-gateway/egress-gateway.rst:
```diff
@@ -104,7 +104,9 @@ The egress gateway feature and all the requirements can be enabled as follow:
-            enable-ipv4-egress-gateway: true
+            enable-egress-gateway: true  # Enables egress gateway for both IPv4 and IPv6
+            # Alternatively, you can use the deprecated flag for IPv4 only:
+            # enable-ipv4-egress-gateway: true
```
If the other option is deprecated, then we shouldn't encourage people to use it.
```suggestion
```

joestringer commented on Documentation/network/egress-gateway/egress-gateway.rst:
```diff
@@ -268,14 +270,25 @@ There are 3 different ways this can be achieved:
+       nodeSelector:
+         matchLabels:
+           testLabel: testVal
+       egressIP: 2001:db8::1  # IPv6 example
```
I don't think we need to instruct users how to specify an IPv6 address. The text above declares that IPv6 is supported, and if the user wants to use IPv6 then they will know how an IPv6 address looks. It doesn't look like there's any special required syntax like wrapping it in speech marks or anything, so we can just keep the existing example.

```suggestion
       egressIP: a.b.c.d
```

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -672,9 +672,12 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	option.BindEnv(vp, option.EnableIPMasqAgent)
 
-	flags.Bool(option.EnableIPv4EgressGateway, false, "Enable egress gateway for IPv4")
+	flags.Bool(option.EnableIPv4EgressGateway, false, "Enable egress gateway for IPv4 (deprecated, use --enable-egress-gateway instead)")
```
```suggestion
	flags.Bool(option.EnableIPv4EgressGateway, false, "Enable egress gateway for IPv4")
	flags.MarkDeprecated(option.EnableIPv4EgressGateway, "use --" + option.EnableEgressGateway)
```

joestringer commented on examples/kubernetes-egress-gateway/egress-gateway-policy.yaml:
```diff
@@ -15,22 +15,27 @@ spec:
   destinationCIDRs:
-  - 192.168.60.13/32
+  - 192.168.60.13/32  # IPv4 example
+  # For IPv6 destinations, you can specify an IPv6 CIDR:
```
It's enough to just include the IPv6 example.
```suggestion
```

joestringer commented on examples/kubernetes-egress-gateway/egress-gateway-policy.yaml:
```diff
@@ -15,22 +15,27 @@ spec:
     # IP used to masquerade traffic leaving the cluster
-    egressIP: "192.168.60.100"
+    egressIP: "192.168.60.100"  # IPv4 example
+    # For IPv6 destinations, you can specify an IPv6 address:
```
```suggestion
```

joestringer reviewed (changes requested):
Thanks for the PR! Some minor feedback below.

You will also need to update the docs pages; the CI workflow for documentation highlights this. To fix, run `make -C Documentation update-cmdref`.

syedazeez337 committed:
Merge branch 'cilium:main' into main

syedazeez337 committed:
egress-gateway: Add IPv6 support

- Add IPv6 support for egress gateway
- Properly mark the old flag as deprecated using flags.MarkDeprecated
- Update documentation to remove unnecessary IPv6 examples and comments
- Update command reference documentation
- Simplify example file

Signed-off-by: Syed Azeez <syedazeez337@gmail.com>
